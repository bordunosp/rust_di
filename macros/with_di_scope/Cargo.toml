[package]
name = "with_di_scope"
version = "0.1.0"
edition = "2024"

[lib]
proc-macro = true

[dependencies]
#proc-macro2 = "1.0.95"
quote = "1.0.40"
syn = { version = "2.0.102", features = ["full", "extra-traits"] }

[dev-dependencies]
trybuild = "1.0.105"
tokio = { version = "1", features = ["full"] }
dashmap = "7.0.0-rc2"
di = { path = "../../", features = ["test_utils"] }
