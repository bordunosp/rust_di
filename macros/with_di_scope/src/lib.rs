use proc_macro::TokenStream;
use quote::quote;
use syn::{ItemFn, parse_macro_input};

#[proc_macro_attribute]
pub fn with_di_scope(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input_fn = parse_macro_input!(item as ItemFn);

    let fn_name = &input_fn.sig.ident;
    let fn_inputs = &input_fn.sig.inputs;
    let fn_output = &input_fn.sig.output;
    let fn_block = &input_fn.block;
    let fn_attrs = &input_fn.attrs;
    let fn_vis = &input_fn.vis;
    let fn_async = &input_fn.sig.asyncness;

    if fn_async.is_none() {
        return syn::Error::new_spanned(
            input_fn.sig.fn_token,
            "The 'with_di_scope' macro can only be applied to async functions.",
        )
        .to_compile_error()
        .into();
    }

    let expanded = quote! {
        #(#fn_attrs)*
        #fn_vis #fn_async fn #fn_name(#fn_inputs) #fn_output {
            di::DIScope::run_with_scope(|| async {
                #fn_block
            }).await
        }
    };

    expanded.into()
}
