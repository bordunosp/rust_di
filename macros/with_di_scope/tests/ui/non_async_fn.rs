use with_di_scope::with_di_scope;

#[with_di_scope]
fn sync_function() {
    println!("This should not compile");
}

fn main() {}
