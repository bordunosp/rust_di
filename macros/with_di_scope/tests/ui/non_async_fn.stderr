error: The 'with_di_scope' macro can only be applied to async functions.
 --> tests/ui/non_async_fn.rs
  |
  | fn sync_function() {
  | ^^
